chain/jsonrpc-adversarial-primitives/Cargo.toml

```diff
@@ -0,0 +1,13 @@
+[package]
+name = "near-jsonrpc-adversarial-primitives"
+version = "0.2.0"
+authors = ["Near Inc <hello@nearprotocol.com>"]
+publish = false
+edition = "2018"
+
+[dependencies]
+serde = { version = "1", features = ["derive"] }
+serde_json = "1"
+near-primitives = { path = "../../core/primitives" }
+near-jsonrpc-primitives = { path = "../jsonrpc-primitives", features = ["adversarial"] }
```

Comment:
Please remove, `features = ["adversarial"]`.
This causes `near-jsonrpc-primitives ` and `near-network` to be build with `adversarial` even if it's not specified.

At the end of the file please add
```
[features]
adversarial = ["near-jsonrpc-primitives/adversarial", "near-network/adversarial"]
```
This will propagate `adversarial` to those 2 crates.

Reply:
> @pmnoxx, please could you take a look at CI.. I tried mapping out the relationship but everything seems to checkout in my head. For some reason, when compiling `near-client` without adversarial features, it's dependency `near-network` is built with the feature regardless.

See me comment about usage of `adversarial`. That should help.

Reply:
Yeah, I already experimented with that locally, didn't seem to work either.

Reply:
@miraclx It works just fine: https://github.com/near/nearcore/pull/4915/files

Reply:
Thanks!

Reply:
Still have a small nit here:

The issue, though tough to explain, is likely a bug in how cargo resolves dep features.

For a crate A with an optional dependency B depending on C/feat where you only directly depend on C.
Interaction between A and C ends up in conflict, despite B not being selected.

```text
A
├── B { optional = true }
│   └── C { features = ["feat"] }
└── C
```

Cargo somehow infers that `A > C { features = ["feat"] }` despite that only being the case if `A { features = ["B"] }`

Now, this patch explicitly tells cargo to include `C/feat` if and only if `B` selected

It's unintuitive and perhaps even a mouthful

---

I started going this route on my own but stopped halfway because it didn't make much sense to me

The crate name is `near-jsonrpc-adversarial-primitives` .. It's expected to have `adversarial` features by default. But now, we have to do

```toml
near-jsonrpc-adversarial-features = { path = .., features = ["adversarial"] }
```

which isn't neat.. at all